core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -641,14 +642,25 @@ public UnassignedIterator iterator() {
          * Should be used with caution, typically,
          * the correct usage is to removeAndIgnore from the iterator.
          * @see #ignored()
-         * @see UnassignedIterator#removeAndIgnore()
+         * @see UnassignedIterator#removeAndIgnore(UnsuccessfulAllocationAttempt)
          * @see #isIgnoredEmpty()
+         * @return true iff the decision caused a change to the unassigned info
          */
-        public void ignoreShard(ShardRouting shard) {
+        public boolean ignoreShard(ShardRouting shard, UnsuccessfulAllocationAttempt allocationAttempt) {
+            boolean changed = false;
             if (shard.primary()) {
                 ignoredPrimaries++;
+                assert shard.unassignedInfo() != null;
+                UnassignedInfo oldUnassignedInfo = shard.unassignedInfo();
```

Comment:
I think it's simpler to compare `shard.unassignedInfo.getLastAllocationAttempt` directly to `allocationAttempt` and create a new UnassignedInfo if it differs.